subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1895,3 +1877,106 @@ int ztls_setsockopt(int sock, int level, int optname,
 
 	return 0;
 }
+
+static ssize_t tls_sock_read_vmeth(void *obj, void *buffer, size_t count)
+{
+	return ztls_recvfrom_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static ssize_t tls_sock_write_vmeth(void *obj, const void *buffer,
+				    size_t count)
+{
+	return ztls_sendto_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static int tls_sock_ioctl_vmeth(void *obj, unsigned int request, ...)
+{
+	switch (request) {
+	case ZFD_IOCTL_CLOSE:
+		return ztls_close_ctx(obj);
+
+	case ZFD_IOCTL_FCNTL: {
+		va_list args;
+		int err;
+
+		/* Pass the call to the core socket implementation. */
+		va_start(args, request);
+		err = sock_fd_op_vtable.fd_vtable.ioctl(obj, request, args);
+		va_end(args);
+
+		return err;
+	}
+
+	default:
+		errno = EOPNOTSUPP;
+		return -1;
+	}
+}
+
+static int tls_sock_bind_vmeth(void *obj, const struct sockaddr *addr,
+			       socklen_t addrlen)
+{
+	return sock_fd_op_vtable.bind(obj, addr, addrlen);
+}
+
+static int tls_sock_connect_vmeth(void *obj, const struct sockaddr *addr,
+				  socklen_t addrlen)
+{
+	return ztls_connect_ctx(obj, addr, addrlen);
+}
+
+static int tls_sock_listen_vmeth(void *obj, int backlog)
+{
+	return sock_fd_op_vtable.listen(obj, backlog);
+}
+
+static int tls_sock_accept_vmeth(void *obj, struct sockaddr *addr,
+				 socklen_t *addrlen)
+{
+	return ztls_accept_ctx(obj, addr, addrlen);
+}
+
+static ssize_t tls_sock_sendto_vmeth(void *obj, const void *buf, size_t len,
+				     int flags,
+				     const struct sockaddr *dest_addr,
+				     socklen_t addrlen)
+{
+	return ztls_sendto_ctx(obj, buf, len, flags, dest_addr, addrlen);
+}
+
+static ssize_t tls_sock_recvfrom_vmeth(void *obj, void *buf, size_t max_len,
+				       int flags, struct sockaddr *src_addr,
+				       socklen_t *addrlen)
+{
+	return ztls_recvfrom_ctx(obj, buf, max_len, flags,
+				 src_addr, addrlen);
+}
+
+static int tls_sock_getsockopt_vmeth(void *obj, int level, int optname,
+				     void *optval, socklen_t *optlen)
+{
+	return ztls_getsockopt_ctx(obj, level, optname, optval, optlen);
+}
+
+static int tls_sock_setsockopt_vmeth(void *obj, int level, int optname,
+				     const void *optval, socklen_t optlen)
+{
+	return ztls_setsockopt_ctx(obj, level, optname, optval, optlen);
+}
+
+
+static const struct socket_op_vtable tls_sock_fd_op_vtable = {
+	.fd_vtable = {
+		.read = tls_sock_read_vmeth,
+		.write = tls_sock_write_vmeth,
+		.ioctl = tls_sock_ioctl_vmeth,
+	},
+	.bind = tls_sock_bind_vmeth,
+	.connect = tls_sock_connect_vmeth,
```

Comment:
Why do we use extra helpers here? For example why not set `.connect = ztls_connect_ctx` directly?

Reply:
Never mind, I saw @pfalcon asked already the same question.